src/gdb/gdb.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2018, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ * GDB Stub - parse incoming GDB packets, control FW accordingly and provide
+ * a reply to the GDB server.
+ *
+ */
+
+#include <sof/gdb/gdb.h>
+#include <sof/gdb/ringbuffer.h>
+#include <arch/gdb/xtensa-defs.h>
+#include <string.h>
+#include <signal.h>
+#include <stdint.h>
+#include <xtensa/xtruntime.h>
+#include <sof/interrupt.h>
+
+/* main buffers */
+static unsigned char remcomInBuffer[BUFMAX];
+static unsigned char remcomOutBuffer[BUFMAX];
+
+/* local functions */
+static int getHex(unsigned char ch);
```

Comment:
coding convention is get_hex, i.e. all lowercase for functions with _ to separate words.

Reply:
>One general question is that we have lots of while(1) {} loops reading or parsing GDB data. What happens if we have some corruption in the GDB command stream (or GDB unexpectedly quits), we could loop forever. We probably need to look at breaking out of the loops on error conditions.

Okay, this is good point actually, so let me elaborate a bit. 

**Corrupted data** - as you can see in code `/* TODO: handle wrong checksums */` there is no support for wrong checksum **on the gdb_stub side** at the moment. However, there is on the server side - it is tcp ip communication between GDB and gdb_server so the incoming data for gdb_stub on the FW side is quite consistent. 
BTW: I've been testing GDB for a weeks now,  guess how many corrupted packet I received? Zero. That's why I didn't bother to implement checksum validation on FW side yet.

**Infinite loop** - yes we constantly monitor incoming packets and once detected valid one we pass it further to parsing stage and this is what we want. What happens when GDB is hung or you lost connection? Then all you need to do is: restart the session and start over, gdb stub should still monitor incoming buffer and accept our requests.

Is that sufficient for you? I think at some point we should organize a video session showing GDB in action.